test/framework/src/main/java/org/elasticsearch/index/mapper/MappingLookupUtils.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.elasticsearch.index.analysis.NamedAnalyzer;
+import org.elasticsearch.index.mapper.FieldMapper.CopyTo;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class MappingLookupUtils {
+    public static MappingLookup fromTypes(MappedFieldType... types) {
+        return fromTypes(Arrays.stream(types));
+    }
+    
+    public static MappingLookup fromTypes(Stream<MappedFieldType> types) {
+        List<FieldMapper> mappers = types.map(MappingLookupUtils::mockFieldMapper).collect(toList());
+        //  Alias <name>-alias to <name> so we can test aliases
+        return new MappingLookup(
+            "_doc",
+            mappers,
+            List.of(),
+            List.of(),
+            List.of(),
+            0,
+            souceToParse -> null,
+            true
+        );
+    }
+
+    public static FieldMapper mockFieldMapper(MappedFieldType type) {
```

Comment:
The way that this is used looks suspicious to me. I see it being used with runtime fields, which don't have a corresponding mapper normally. With this method they end up being exposed as ordinary fields to tests?

Reply:
There should be a separate code-path for creating runtime fields, that only provides them to MappingLookup without them causing the creation of field mappers.

Reply:
`QueryShardContextTests` has been doing that sort of thing for a month or so now. We didn't have runtime fields without the `MappedFieldType` back then. I've pushed 83f9f30acfe3557643603d00444fd05ff48a1c42 to fork off runtime fields. I'm not a huge fan of the `instanceof` that it adds but it seems like a useful thing for testing.